datalad/runner/runnerthreads.py

```diff
@@ -0,0 +1,298 @@
+import logging
+import os
+import threading
+from abc import (
+    abstractmethod,
+    ABCMeta,
+)
+from enum import Enum
+from queue import (
+    Empty,
+    Full,
+    Queue,
+)
+from typing import (
+    Any,
+    IO,
+    List,
+    Optional,
+    Tuple,
+    Union,
+)
+
+
+lgr = logging.getLogger("datalad.runner.runnerthreads")
+
+
+class IOState(Enum):
+    ok = "ok"
+    timeout = "timeout"
+
+
+class ExitingThread(threading.Thread):
+    def __init__(self):
+        super().__init__(daemon=True)
+        self.exit_requested = False
+
+    def request_exit(self):
+        """
+        Request the thread to exit. This is not guaranteed to
+        have any effect, because the instance has to check for
+        self.exit_requested and act accordingly. It might not
+        do that.
+        """
+        self.exit_requested = True
+
+
+class BlockingOSReaderThread(ExitingThread):
+    """
+    A blocking OS file reader. If it reads
+    anything, it stores its data in a queue.
+    That allows a consumer to block
+    on OS-reads with timeout, independent from
+    the OS read capabilities.
+    It enqueues bytes if something is read. If
+    the file is close, it will enqueue None and
+    exit.
```

Comment:
should it rhyme if so narrow? ;)

Reply:
You mean something like this: 

    It enqueues bytes if something's read.
    If source is closed, it ends the thread,
    not without enqueuing None,
    so you might see that it is done.

;)

Reply:
LOL! 😎